libs/langchain/langchain/vectorstores/neo4j_vector.py

```diff
@@ -0,0 +1,623 @@
+        self.embedding_node_property = embedding_node_property
+        self.text_node_property = text_node_property
+        self.logger = logger or logging.getLogger(__name__)
+        self.override_relevance_score_fn = relevance_score_fn
+
+        # Calculate embedding dimension
+        self.embedding_dimension = len(embedding_function.embed_query("foo"))
+
+        # Delete existing data if flagged
+        if pre_delete_collection:
+            self.query(f"MATCH (n:{self.node_label}) DETACH DELETE n")
+
+    def query(self, query: str, params: dict = {}) -> List[Dict[str, Any]]:
+        """
+        This method sends a Cypher query to the connected Neo4j database
+        and returns the results as a list of dictionaries.
+
+        Args:
+            query (str): The Cypher query to execute.
+            params (dict, optional): Dictionary of query parameters. Defaults to {}.
+
+        Returns:
+            List[Dict[str, Any]]: List of dictionaries containing the query results.
+
+        Raises:
+            ValueError: If the generated Cypher statement is not valid.
+        """
+        from neo4j.exceptions import CypherSyntaxError
+
+        with self._driver.session(database=self._database) as session:
+            try:
+                data = session.run(query, params)
+                return [r.data() for r in data]
+            except CypherSyntaxError as e:
+                raise ValueError(f"Generated Cypher Statement is not valid\n{e}")
+
+    def verify_version(self) -> None:
+        """
+        Check if the connected Neo4j database version supports vector indexing.
+
+        Queries the Neo4j database to retrieve its version and compares it
+        against a target version (5.11.0) that is known to support vector
+        indexing. Raises a ValueError if the connected Neo4j version is
+        not supported.
+        """
+        version = self.query("CALL dbms.components()")[0]["versions"][0]
+        version_tuple = tuple(map(int, version.split(".")))
+        target_version = (5, 11, 0)
+
+        if version_tuple < target_version:
+            raise ValueError(
+                "Version index is only supported in Neo4j version 5.11 or greater"
+            )
+
+    def retrieve_existing_index(self) -> Optional[int]:
+        """
+        Check if the vector index exists in the Neo4j database
+        and returns its embedding dimension.
+
+        This method queries the Neo4j database for existing indexes
+        and attempts to retrieve the dimension of the vector index
+        with the specified name. If the index exists, its dimension is returned.
+        If the index doesn't exist, `None` is returned.
+
+        Returns:
+            int or None: The embedding dimension of the existing index if found.
+        """
+
+        index_information = self.query(
+            "SHOW INDEXES YIELD name, type, labelsOrTypes, properties, options "
+            "WHERE type = 'VECTOR' AND name = $index_name "
```

Comment:
perhaps better to not rely on the name but check the label + property instead? as there could be a conflicting index

Reply:
Yeah, we can definitely do that... maybe do both